core/modules/menu/menu.module

```diff
@@ -856,3 +856,85 @@ function menu_preprocess_block(&$variables) {
     $variables['attributes']['role'] = 'navigation';
   }
 }
+
+/*
+ * @todo In theory we should take over cache_menu and re-act to cache clear
+ *       events by invalidating a menu_view tag.
+ */
+
+/**
+ * Implements hook_preprocess_menu_link().
+ */
+function menu_preprocess_menu_link(&$variables) {
+  $link = $variables['element']['#original_link'];
+
+  // Unfortunately preprocess_menu_tree does not have the information we need,
+  // so we need to add the cache tag for each menu link in a menu separately.
+  backdrop_add_cache_tags(_menu_menu_link_get_cache_tags($link));
+}
+
+/**
+ * Implements hook_preprocess_links().
+ */
+
+function menu_preprocess_links(&$variables) {
+  $tag_map = array();
+  if (isset($variables['links'])) {
+    foreach ($variables['links'] as $link) {
+      if (isset($link['#menu_name'])) {
+        $tag_map['menu:' . strtr($link['#menu_name'], '-', '_')] = TRUE;
+      }
+    }
+    backdrop_add_cache_tags(array_keys($tag_map));
+  }
+}
+
+/**
+ * Implements hook_menu_link_alter().
+ */
+function menu_menu_link_alter(&$item) {
+  // @todo This needs an upgrade path for existing menu items.
+  $item['options']['#menu_name'] = $item['menu_name'];
+}
+
+/**
+ * Implements hook_menu_link_insert().
+ */
+function menu_menu_link_insert($link) {
+  backdrop_invalidate_cache_tags(_menu_menu_link_get_cache_tags($link));
+}
+
+/**
+ * Implements hook_menu_link_update().
+ */
+function menu_menu_link_update($link) {
+  backdrop_invalidate_cache_tags(_menu_menu_link_get_cache_tags($link));
+}
+
+/**
+ * Implements hook_menu_item_delete().
+ */
+function menu_menu_link_delete($link) {
+  backdrop_invalidate_cache_tags(_menu_menu_link_get_cache_tags($link));
+}
+
+/* -----------------------------------------------------------------------
```

Comment:
It's not clear what these placeholders are for.

Reply:
Those were leftover from copying the code from the D7 module. I just forgot to remove them from menu.module